### PR TITLE
Perf improvements in CCW creation.

### LINKED ISSUE #438: Improve perf of CreateCCWForObject by caching.
WinRT.ComWrappersSupport.CreateCCWForObject(object) shows up in DP perf investigation traces.  The main cost is associated with calculating VTables (this seems to be partially due to fetching custom attributes).  Ideally (if possible), caching of VTables for type could speed it up.  If that's not possible (due to these changing at runtime), caching interface ptrs would reduce the number of calls to get custom attributes and would still give an improvement.

### LINKED ISSUE #554: VTable memory should be handled by Type not Object
The following table:

https://github.com/microsoft/CsWinRT/blob/fab6067c83e9063a8c0eda387b4c250ca5a4da62/WinRT.Runtime/ComWrappersSupport.net5.cs#L161

stores memory for vtables on the `object` level instead of the `Type` level. This is creating an enormous amount of unnecessary memory pressure on the application. Converting this to be on the `Type` rather than `object` would make CCW creation faster and reduce memory usage.

/cc @jkoritzinsky

### PR DESCRIPTION
**Pull request by manodasanW**
- Caching the vtable entries per type and as part of that also moving vtable cleanup to per type rather than per object.
- Also moving to store IInspectableInfo per type rather than per object.

Based on a perf and memory analysis, it can be seen the memory consumption for storing the vtable entries is down as expected and in the repro scenario where CCW creation was found to be expensive, it is no longer near the top of the CPU usage and the execution time of the repro seems to have went down by half.

Fixes #438 
Fixes #554

### CODE REVIEW

**Comment by AaronRobinsonMSFT** [comment]
FYI @jkotas

**Review by Scottj1s — APPROVED** [review]

File: src/WinRT.Runtime/ComWrappersSupport.net5.cs [hunk 1]
```diff
@@ -212,51 +212,57 @@ static unsafe DefaultComWrappers()
...
+                });
+
+                // This should be the last entry as it is included / excluded based on the flags.
+                entries.Add(new ComInterfaceEntry
+                {
+                    IID = typeof(IUnknownVftbl).GUID,
+                    Vtable = IUnknownVftbl.AbiToProjectionVftblPtr
+                });
+
+                ComInterfaceEntry* nativeEntries = (ComInterfaceEntry*)Marshal.AllocCoTaskMem(sizeof(ComInterfaceEntry) * entries.Count);
```
Comment (AaronRobinsonMSFT) [inline src/WinRT.Runtime/ComWrappersSupport.net5.cs:239]: I would avoid the manual managed allocation here and use [`RuntimeHelpers.AllocateTypeAssociatedMemory()`](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.runtimehelpers.allocatetypeassociatedmemory) instead.

Comment (manodasanW) [reply src/WinRT.Runtime/ComWrappersSupport.net5.cs:239]: I did look at using it, but I didn't have a static field associated with the type that I could use to store it and I was concerned about storing this address in the ConditionalWeakTable as I wasn't sure whether there is a scenario where the memory allocated from this function might be cleaned up before the entry is removed / invalidated from the weak table.  If you think this latter case won't happen (i.e. I can't retrieve an entry from a ConditionalWeakTable based on type for which the type had already executed FreeMemory for any type associated memory), I can move to using it.

Comment (AaronRobinsonMSFT) [reply src/WinRT.Runtime/ComWrappersSupport.net5.cs:239]: >i.e. I can't retrieve an entry from a ConditionalWeakTable based on type for which the type had already executed FreeMemory for any type associated memory

Yep. This function was designed for exactly the scenario being performed here. The fix is simple - change the above call and remove the finalizer doing the cleanup in `VtableEntries`.

Comment (AaronRobinsonMSFT) [reply src/WinRT.Runtime/ComWrappersSupport.net5.cs:239]: This will also reduce pressure on the finalizer queue.

**Comment by AaronRobinsonMSFT** [comment]
>Based on a perf and memory analysis, it can be seen the memory consumption for storing the vtable entries is down as expected and in the repro scenario where CCW creation was found to be expensive, it is no longer near the top of the CPU usage and the execution time of the repro seems to have went down by half.

@manodasanW I appreciate the following request may be annoying, but would it be possible to get a sense of the raw numbers we are looking at here? Like Scenario A vs Scenario B and then % improvement? Memory used and CPU usage are all that is interesting at the moment. It helps the .NET team talk about the investments we make in APIs and scenarios they unlock.

**Review by BenJKuhn — APPROVED** [review]

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by manodasanW** [comment]
> > Based on a perf and memory analysis, it can be seen the memory consumption for storing the vtable entries is down as expected and in the repro scenario where CCW creation was found to be expensive, it is no longer near the top of the CPU usage and the execution time of the repro seems to have went down by half.
> 
> @manodasanW I appreciate the following request may be annoying, but would it be possible to get a sense of the raw numbers we are looking at here? Like Scenario A vs Scenario B and then % improvement? Memory used and CPU usage are all that is interesting at the moment. It helps the .NET team talk about the investments we make in APIs and scenarios they unlock.

I plan on asking Bartek if he can do an updated run of this scenario which he has been tracking for various CsWinRT versions and the previous built-in version.  I will also look at getting some metrics for memory usage.

**Comment by AaronRobinsonMSFT** [comment]
>I plan on asking Bartek if he can do an updated run of this scenario which he has been tracking for various CsWinRT versions and the previous built-in version. I will also look at getting some metrics for memory usage.

Perfect. That would be most insightful. Thanks!

**Pull request merged**